### ISSUE TITLE
Treat liveness probe-based restarts as a voluntary disruptions gated by PodDisruptionBudgets

### ISSUE THREAD

**Issue by MarcPow** [kubernetes/kubernetes#123204]
### What would you like to be added?

Prevent liveness probes from driving all pods in a cluster down (by treating a container restart as a gated, voluntary disruption)

### Why is this needed?

Overly sensitive liveness probes have the potential to cause outage.  Categorizing restarts as voluntary disruptions that must be budget checked provides natural gating to protect Service availability.

**Comment by MarcPow** [comment]
This appears related to #66230.  The objections therein seem to be related to implementation complexity, but it's not clear to me why we can't simply treat a container restart as a pseudo-eviction and get the behavior that we want.  No one can argue that an interior pod restart is semantically, if not physically, not equivalent to a pod eviction followed by a rescheduling on the same node.

Perhaps we could solve this in sig-node by adding a liveness-failure-mode that actually promotes to evicting the pod itself (rather than restarting the container).  If that opt-in failure mode were added, we'd get PDB checks for free.

**Comment by thockin** [comment]
@tallclair

**Comment by sftim** [comment]
Restarting a container is **not** an eviction, not even a pseudo one. If we were failing Pods outright (which Kubernetes AIUI right now doesn't do as a direct outcome from a probe failure), then a disruption budget might be appropriate.

Bear in mind that container restarts on probe failure are available to static Pods; this is super useful for keeping the control plane available in the event of a partial outage. It can even help with a situation where all the API server Pods fail, but in a way that they can recover from individually.

> Perhaps we could solve this in sig-node by adding a liveness-failure-mode that actually promotes to evicting the pod itself (rather than restarting the container). If that opt-in failure mode were added, we'd get PDB checks for free.

Yes, that feels like the way to go.

---

If we still want to support conditional container restarts, I would add a new opt-in API. The kubelet would try to create or interact with a subresource on Pod. For Pods that opted in, the kubelet would respond to container probe failure by trying incrementing the container restart request counter. And then if that change to status was admitted, the kubelet increments the container restart counter.

```yaml
spec:
  ...
  containers:
  - name: example
    image: registry.k8s.io/pause:1-healthcheckable
    args:
    - --forever
    livenessProbe:
      failureThreshold: 5
      httpGet:
        path: /health
        port: 8080
        scheme: HTTP
      initialDelaySeconds: 60
      probeFailureAction: container-restart-request # new field, new value
      periodSeconds: 10
      successThreshold: 1
      timeoutSeconds: 5
    restartRequestCount: 18                         # exposed via a subresource?
status:
   ...
  containerStatuses:
  - name: example
    ready: true
    containerID: 8d6a0af603edb6c668967d64a2a6
    ...
    restartCount: 23                                # how many actual restarts
```

**Comment by MarcPow** [comment]
>> Perhaps we could solve this in sig-node by adding a liveness-failure-mode that actually promotes to evicting the pod itself (rather than restarting the container). If that opt-in failure mode were added, we'd get PDB checks for free.
> Yes, that feels like the way to go.

Modifying your example, perhaps we can simplify:

```yaml
containers:
  - name: example
    image: registry.k8s.io/pause:1-healthcheckable
    args:
    - --forever
    livenessProbe:
      failureThreshold: 5
      httpGet:
        path: /health
        port: 8080
        scheme: HTTP
      initialDelaySeconds: 60
      probeFailureAction: evict-containing-pod
      periodSeconds: 10
      successThreshold: 1
      timeoutSeconds: 5
```

**Comment by thockin** [comment]
Are you asking for a liveness failure to evict a pod from a node because that's what you want, or are you just trying to trick someone into checking the PDB?

**Comment by sftim** [comment]
For my part, I'd like the option to fail a whole Pod on a liveness check failure. Maybe the thing that needs to happen is to re-run the init containers because some config has changed and the Pod isn't quite as cloud-native as it could be. 

Replacing the Pod feels like the right thing there; on the other hand, it shouldn't be the default choice. We should default to assuming that apps are reasonably cloud-native, and encouraging (nudging) people to make them more cloud native when a situation like this example crops up.

**Comment by sftim** [comment]
> ```
>       probeFailureAction: evict-containing-pod
> ```

This is a helpful thing to be offering, but I don't see how it respects PodDisruptionBudgets; I am wary of suggesting that the kubelet somehow adds more things to track.

**Comment by sftim** [comment]
The simplest option here is to do nothing and encourage people to extend Kubernetes to get the behavior they want. It can be done (for example: watch a Pod for a restart count going up for a container you care about, and then request eviction for the relevant Pod whilst respecting any configured disruption budgets).

This doesn't feel to me like in-tree support would be justified. An in-project helper controller would be useful, and all it needs is a rich seam of contributor time.

**Comment by MarcPow** [comment]
@thockin 

> Are you asking for a liveness failure to evict a pod from a node because that's what you want, or are you just trying to trick someone into checking the PDB?

Hmm.  That's a fair question.   The simple answer is:  There are real-life scenarios where either (a) correlated back-pressure or (b) poison pill traffic might create more simultaneous restarts than I'm comfortable with - and the best way to guard against runaway repair is a budget.

The more complex answer is: Probe failure induced reactions are semantically (perhaps morally? :) ) voluntary disruptions, for all that they're self-induced.  The idea that they're not checked against the budget is a miss.   [Of course, I realize that I don't get to re-define semantic categories just because I'd like to - and they weren't in the original spec - but I'm going to try anyway.  I think extending the definition here has the potential to drive up safety and service quality.]

A secondary answer is:  I've seen probes start to fail because nodes go bad.   Being able to say "If you see this particular probe fail, please try a different node" unlocks even more value in those scenarios.

**Comment by MarcPow** [comment]
@sftim 

> The simplest option here is to do nothing and encourage people to extend Kubernetes to get the behavior they want. It can be done (for example: watch a Pod for a restart count going up for a container you care about, and then request eviction for the relevant Pod whilst respecting any configured disruption budgets).

I hear you - I've worked on teams that have built such secondary controllers - and they work great, but they only solve for the "secondary answer" I noted to @thockin above.  If you can watch particular container restart counts go up and use it as a criteria to say "Hey man, this node is probably bad.  Let's try another" an external controller works great.   I've seen it be a very effective strategy.

But if your goal is to stop too many containers for a given Deployment restart at once, well - the external controller can't catch that, because it's an ex-post-facto action, rather than a pre-action interceptor.

**Comment by thockin** [comment]
> might create more simultaneous restarts than I'm comfortable with

I ACK this problem, but offer questions in response.

What if those containers are legitimately crash?  You're going to restart anyway.  We used to call that the Query Of Death.  A well-meaning client accidentally triggers a panic/segfault, so the query fails.  The client or the LB sends it again, to a new backend and repeat.

If the LivenessProbe fails, that's *direct* evidence that the pod is in trouble.  It's not responding to even the most rudimentary queries.  For all intents - it's dead.  Now, hypothetically it could recover in the next millisecond, but we don't know that.  You, the definer of the probe, chose that timeout and that failure count.  Why should we need to check beyond that?

Practically speaking, you're sort of describing a token-bucket for pod restarts, right?  Kubelet checks if it can get permission to restart a pod, where the answer is either "yes" or "ask again soon".  We don't have an easy way to do that.

> if your goal is to stop too many containers for a given Deployment restart at once

If you failed a probe, I would argue we're not "stopping" you - you are already stopped.  We're "restarting" you.

> "If you see this particular probe fail, please try a different node"

The problem with that statement is that we don't have an obvious way to poison this node - the pod will be killed and replaced and the scheduler will make an independent decision, possibly choosing the same node.

 > For my part, I'd like the option to fail a whole Pod on a liveness check failure

I think it could be interesting for a pod to define different signals for different reactions.  E.g. "If probe A fails, restart this container.  If probe B fails, do a full pod reboot.  If probe C fails, kill the pod completely."  I could also see it being an escalating "ladder" of config.  E.g. "If probe fails, restart this container but if it fails 10 times in 5 minutes, do a full pod reboot, but if that happens twice, just kill this pod completely".  I could also see it defined in terms of exit codes.  That's all pretty complicated - even more if you combine it with a token bucket.  I am not convinced we have the justification for anything like this (just intuition).

TL;DR: I'm still not sure what the real ask is, in practical terms.

**Comment by MarcPow** [comment]
> If the LivenessProbe fails, that's direct evidence that the pod is in trouble. It's not responding to even the most rudimentary queries. For all intents - it's dead. Now, hypothetically it could recover in the next millisecond, but we don't know that. You, the definer of the probe, chose that timeout and that failure count. Why should we need to check beyond that?

You are pre-supposing that pod authors are effective at determining that "the pods is in trouble" or that "it's dead".  But I've seen too many people, over too many years, be over-confident in their assessments in health monitors and get themselves in trouble.  

The current contract says "I have to fail X times before you consider me dead".  That's a sane model, you're trying to buffer against mis-assessment. "Come on! What's the probability that a pod declares itself dead ten times in a row and gets it wrong?"  But it protects only against transients, not outright mischaracterizations.

But what if a single pod gets it wrong ten times in a row?  Well then, you restart a single pod - it's not a big deal. But if all the pods of the same type get it wrong in the same way at the same time?  Then, well, you're not having a very good day.

> TL;DR: I'm still not sure what the real ask is, in practical terms.

I'd like to be able to set 'repairMode: attemptEviction' on a livenessProbe.  In that mode, if the probe crosses the repair threshold, we call the pod eviction api.

**Comment by thockin** [comment]
> You are pre-supposing that pod authors are effective at determining that "the pods is in trouble"

I have to assume the author of the pod knows more about it than I do, right?

> But if all the pods of the same type get it wrong in the same way at the same time? Then, well, you're not having a very good day.

Counter-argument: I'd rather know about it so I can maybe catch it in testing) than have something try to paper over it.

> The original ask was to budget-check restarts. I take your point that that's difficult to achieve.

It's not that it's super technically hard, but a) we don't have a mechanism (we have no self-refilling token-bucket API); b) I don't know if we have precedent for something like this (leases, maybe?); c) we end up with another config that has to be correct, in another API.  If users can't get liveness probes right, what makes us think they would get this right?

>  I'd like to be able to set 'repairMode: attemptEviction' on a livenessProbe. In that mode, if the probe crosses the repair threshold, we call the pod eviction api.

I'm not against an escalating response to repeatedly-failing probes, but ultimately that is a hyper-local decision, which I am not sure would give you what you want.  I'm certainly not against someone trying the idea out to see how well it fits.  Just be careful of throwing the baby out with the bathwater - it seems like this could make things worse in an attempt to make them better.

**Comment by thockin** [comment]
> Is that a "triage accepted"? 

Only if the title is reworded.

Something like "Provide better aggregate protection for groups of pods with bad probes" ?

**Comment by tallclair** [comment]
Don't the eviction semantics just describe a `restart: Never` pod? If not, what's the difference?

I think I need to pile on with the other Tim's here, and say I'd prefer to see something like this built out of tree, at least to start. Perhaps a controller that issues the probe itself and handles the restarts. I think the piece we're missing there is a way to remotely trigger a container restart, which I could see an argument for adding.

**Comment by MarcPow** [comment]
@thockin / @tallclair - To close the loop, you folks convinced me to build something out of tree first, and I'll circle back with data on how that runs when ready.

**Comment by thockin** [comment]
Awesome.  I look forward to hearing about it!  Could make a great KubeCon talk.
